src/coord/src/persistence.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+
+use log::{debug, error};
+use timely::progress::Antichain;
+use tokio::sync::mpsc;
+
+use dataflow_types::Update;
+use expr::GlobalId;
+use repr::Timestamp;
+use storage::{Compacter, CompacterMessage, Message, Trace, WriteAheadLogs};
+
+#[derive(Clone, Debug)]
+pub struct PersistenceConfig {
+    pub wals_path: PathBuf,
+    pub traces_path: PathBuf,
+}
+
+pub struct PersistentTables {
+    wals: WriteAheadLogs,
+    compacter_tx: mpsc::UnboundedSender<CompacterMessage>,
+    wals_path: PathBuf,
+    traces_path: PathBuf,
+    disabled: bool,
+}
+
+impl PersistentTables {
+    pub fn new(config: &PersistenceConfig) -> Option<Self> {
+        let (compacter_tx, compacter_rx) = mpsc::unbounded_channel();
+        let mut compacter = match Compacter::new(
+            compacter_rx,
+            config.traces_path.clone(),
+            config.wals_path.clone(),
+        ) {
+            Ok(compacter) => compacter,
+            Err(e) => {
+                error!(
+                    "Encountered error while trying to initialize compacter task: {}",
+                    e
+                );
+                return None;
+            }
+        };
+        tokio::spawn(async move { compacter.run().await });
+        let wals = match WriteAheadLogs::new(config.wals_path.clone()) {
+            Ok(wals) => wals,
+            Err(e) => {
+                error!(
+                    "encountered error while trying to initialize write-ahead log: {}",
+                    e
+                );
+                return None;
+            }
+        };
+
+        Some(Self {
+            wals,
+            compacter_tx,
+            wals_path: config.wals_path.clone(),
+            traces_path: config.traces_path.clone(),
+            disabled: false,
+        })
+    }
+
+    pub fn resume(&mut self, id: GlobalId) -> Option<Vec<Message>> {
+        if self.disabled {
+            return None;
+        }
+
+        let persisted_trace = match Trace::resume(id, &self.traces_path, &self.wals_path) {
+            Ok(trace) => trace,
+            Err(e) => {
+                error!(
+                    "encountered error while trying to restart table {} {}",
```

Comment:
If a table was created on MZ without ```--persistent-tables``` and then MZ is restarted with ```--persistent-tables```, an error is thrown immediately on startup:

```
2021-03-02T10:13:04.006003Z ERROR coord::persistence: encountered error while trying to restart table u1 failed to read mzdata/table_traces/u1 looking for ^batch-[0-9]+-[0-9]+$    
```

Persistence is borked forever until the offending table is dropped.

Reply:
yeah ideally we want the `persistent-tables` flag to be sticky just like `--experimental` is. however, I think it makes more sense to try to test out persistence / fix enough durability flags that we feel comfortable getting rid of the `--persistent-tables` flag and just running with persistence enabled in `--experimental` at which point this would no longer be an issue. I think thats achievable in the 2 - 3 week horizon. I've also upgraded all the error messages to be more clear that persistence is now permanently disabled.